From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Allen Sun <allen.haha@hotmail.com>
Date: Fri, 26 Jun 2026 20:00:04 +0800
Subject: [PATCH] Make entity collision checks less strict


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index e864a1efdc7c852983b71cb614202f5e9550873f..f8481781bf614d8ca75dbb5bb3bfb3c1a3b90a39 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1558,78 +1558,37 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         return list.isEmpty() ? distance : -Shapes.collide(Direction.Axis.Y, boundingBox, list, -distance);
     }
 
-    // Paper start - optimise collisions
-    public Vec3 collide(Vec3 movement) { // Leaves - private -> public
-        final boolean xZero = movement.x == 0.0;
-        final boolean yZero = movement.y == 0.0;
-        final boolean zZero = movement.z == 0.0;
-        if (xZero & yZero & zZero) {
-            return movement;
-        }
-
-        final AABB currentBox = this.getBoundingBox();
-
-        final List<VoxelShape> potentialCollisionsVoxel = new ArrayList<>();
-        final List<AABB> potentialCollisionsBB = new ArrayList<>();
-
-        final AABB initialCollisionBox;
-        if (xZero & zZero) {
-            // note: xZero & zZero -> collision on x/z == 0 -> no step height calculation
-            // this specifically optimises entities standing still
-            initialCollisionBox = movement.y < 0.0 ?
-                ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.cutDownwards(currentBox, movement.y) : ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.cutUpwards(currentBox, movement.y);
-        } else {
-            initialCollisionBox = currentBox.expandTowards(movement);
-        }
-
-        final List<AABB> entityAABBs = new ArrayList<>();
-        ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.getEntityHardCollisions(
-            this.level, (Entity)(Object)this, initialCollisionBox, entityAABBs, 0, null
-        );
-
-        ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.getCollisionsForBlocksOrWorldBorder(
-            this.level, (Entity)(Object)this, initialCollisionBox, potentialCollisionsVoxel, potentialCollisionsBB,
-            ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_CHECK_BORDER, null
-        );
-        potentialCollisionsBB.addAll(entityAABBs);
-        final Vec3 collided = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.performCollisions(movement, currentBox, potentialCollisionsVoxel, potentialCollisionsBB);
-
-        final boolean collidedX = collided.x != movement.x;
-        final boolean collidedY = collided.y != movement.y;
-        final boolean collidedZ = collided.z != movement.z;
-
-        final boolean collidedDownwards = collidedY && movement.y < 0.0;
-
-        final double stepHeight;
-
-        if ((!collidedDownwards && !this.onGround) || (!collidedX && !collidedZ) || (stepHeight = (double)this.maxUpStep()) <= 0.0) {
-            return collided;
-        }
-
-        final AABB collidedYBox = collidedDownwards ? currentBox.move(0.0, collided.y, 0.0) : currentBox;
-        AABB stepRetrievalBox = collidedYBox.expandTowards(movement.x, stepHeight, movement.z);
-        if (!collidedDownwards) {
-            stepRetrievalBox = stepRetrievalBox.expandTowards(0.0, (double)-1.0E-5F, 0.0);
-        }
-
-        final List<VoxelShape> stepVoxels = new ArrayList<>();
-        final List<AABB> stepAABBs = entityAABBs;
+    // Paper start - optimise collisions reverted to vanilla
+    public Vec3 collide(Vec3 vec) { // Leaves - private -> public
+        AABB boundingBox = this.getBoundingBox();
+        List<VoxelShape> entityCollisions = this.level().getEntityCollisions(this, boundingBox.expandTowards(vec));
+        Vec3 vec3 = vec.lengthSqr() == 0.0 ? vec : collideBoundingBox(this, vec, boundingBox, this.level(), entityCollisions);
+        boolean flag = vec.x != vec3.x;
+        boolean flag1 = vec.y != vec3.y;
+        boolean flag2 = vec.z != vec3.z;
+        boolean flag3 = flag1 && vec.y < 0.0;
+        if (this.maxUpStep() > 0.0F && (flag3 || this.onGround()) && (flag || flag2)) {
+            AABB aabb = flag3 ? boundingBox.move(0.0, vec3.y, 0.0) : boundingBox;
+            AABB aabb1 = aabb.expandTowards(vec.x, this.maxUpStep(), vec.z);
+            if (!flag3) {
+                aabb1 = aabb1.expandTowards(0.0, -1.0E-5F, 0.0);
+            }
 
-        ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.getCollisionsForBlocksOrWorldBorder(
-            this.level, (Entity)(Object)this, stepRetrievalBox, stepVoxels, stepAABBs,
-            ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_CHECK_BORDER, null
-        );
+            List<VoxelShape> list = collectColliders(this, this.level, entityCollisions, aabb1);
+            float f = (float)vec3.y;
+            float[] floats = collectCandidateStepUpHeights(aabb, list, this.maxUpStep(), f);
 
-        for (final float step : calculateStepHeights(collidedYBox, stepVoxels, stepAABBs, (float)stepHeight, (float)collided.y)) {
-            final Vec3 stepResult = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.performCollisions(new Vec3(movement.x, (double)step, movement.z), collidedYBox, stepVoxels, stepAABBs);
-            if (stepResult.horizontalDistanceSqr() > collided.horizontalDistanceSqr()) {
-                return stepResult.add(0.0, collidedYBox.minY - currentBox.minY, 0.0);
+            for (float f1 : floats) {
+                Vec3 vec31 = collideWithShapes(new Vec3(vec.x, f1, vec.z), aabb, list);
+                if (vec31.horizontalDistanceSqr() > vec3.horizontalDistanceSqr()) {
+                    double d = boundingBox.minY - aabb.minY;
+                    return vec31.subtract(0.0, d, 0.0);
+                }
             }
         }
-
-        return collided;
-        // Paper end - optimise collisions
+        return vec3;
     }
+    // Paper end - optimise collisions reverted to vanilla
 
     private static float[] collectCandidateStepUpHeights(AABB box, List<VoxelShape> colliders, float deltaY, float maxUpStep) {
         FloatSet set = new FloatArraySet(4);
@@ -1653,8 +1612,28 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     public static Vec3 collideBoundingBox(@Nullable Entity entity, Vec3 vec, AABB collisionBox, Level level, List<VoxelShape> potentialHits) {
-        List<VoxelShape> list = collectColliders(entity, level, potentialHits, collisionBox.expandTowards(vec));
-        return collideWithShapes(vec, collisionBox, list);
+        if (vec.lengthSqr() == 0.0) {
+            return vec;
+        }
+
+        Vec3 vec3 = Vec3.ZERO;
+        for (Direction.Axis axis : Direction.axisStepOrder(vec)) {
+            double d = vec.get(axis);
+            if (d != 0.0) {
+                double expX = (axis == Direction.Axis.X) ? d : 0.0;
+                double expY = (axis == Direction.Axis.Y) ? d : 0.0;
+                double expZ = (axis == Direction.Axis.Z) ? d : 0.0;
+
+                AABB currentBox = collisionBox.move(vec3);
+                AABB queryBox = currentBox.expandTowards(expX, expY, expZ);
+
+                List<VoxelShape> shapes = collectColliders(entity, level, potentialHits, queryBox);
+
+                double d1 = Shapes.collide(axis, currentBox, shapes, d);
+                vec3 = vec3.with(axis, d1);
+            }
+        }
+        return vec3;
     }
 
     public static List<VoxelShape> collectAllColliders(@Nullable Entity entity, Level level, AABB boundingBox) {
@@ -2942,110 +2921,24 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     public boolean isInWall() {
-        // Paper start - optimise collisions
+        // Paper start - optimise collisions reverted to vanilla
         if (this.noPhysics) {
             return false;
-        }
-
-        final double reducedWith = (double)(this.dimensions.width() * 0.8F);
-        final AABB boundingBox = AABB.ofSize(this.getEyePosition(), reducedWith, 1.0E-6D, reducedWith);
-        final Level world = this.level;
-
-        if (ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.isEmpty(boundingBox)) {
-            return false;
-        }
-
-        final int minBlockX = Mth.floor(boundingBox.minX);
-        final int minBlockY = Mth.floor(boundingBox.minY);
-        final int minBlockZ = Mth.floor(boundingBox.minZ);
-
-        final int maxBlockX = Mth.floor(boundingBox.maxX);
-        final int maxBlockY = Mth.floor(boundingBox.maxY);
-        final int maxBlockZ = Mth.floor(boundingBox.maxZ);
-
-        final int minChunkX = minBlockX >> 4;
-        final int minChunkY = minBlockY >> 4;
-        final int minChunkZ = minBlockZ >> 4;
-
-        final int maxChunkX = maxBlockX >> 4;
-        final int maxChunkY = maxBlockY >> 4;
-        final int maxChunkZ = maxBlockZ >> 4;
-
-        final int minSection = ca.spottedleaf.moonrise.common.util.WorldUtil.getMinSection(world);
-        final net.minecraft.world.level.chunk.ChunkSource chunkSource = world.getChunkSource();
-        final BlockPos.MutableBlockPos mutablePos = new BlockPos.MutableBlockPos();
-
-        for (int currChunkZ = minChunkZ; currChunkZ <= maxChunkZ; ++currChunkZ) {
-            for (int currChunkX = minChunkX; currChunkX <= maxChunkX; ++currChunkX) {
-                final net.minecraft.world.level.chunk.LevelChunkSection[] sections = chunkSource.getChunk(currChunkX, currChunkZ, net.minecraft.world.level.chunk.status.ChunkStatus.FULL, true).getSections();
-
-                for (int currChunkY = minChunkY; currChunkY <= maxChunkY; ++currChunkY) {
-                    final int sectionIdx = currChunkY - minSection;
-                    if (sectionIdx < 0 || sectionIdx >= sections.length) {
-                        continue;
-                    }
-                    final net.minecraft.world.level.chunk.LevelChunkSection section = sections[sectionIdx];
-                    if (section.hasOnlyAir()) {
-                        // empty
-                        continue;
-                    }
-
-                    final net.minecraft.world.level.chunk.PalettedContainer<net.minecraft.world.level.block.state.BlockState> blocks = section.states;
-
-                    final int minXIterate = currChunkX == minChunkX ? (minBlockX & 15) : 0;
-                    final int maxXIterate = currChunkX == maxChunkX ? (maxBlockX & 15) : 15;
-                    final int minZIterate = currChunkZ == minChunkZ ? (minBlockZ & 15) : 0;
-                    final int maxZIterate = currChunkZ == maxChunkZ ? (maxBlockZ & 15) : 15;
-                    final int minYIterate = currChunkY == minChunkY ? (minBlockY & 15) : 0;
-                    final int maxYIterate = currChunkY == maxChunkY ? (maxBlockY & 15) : 15;
-
-                    for (int currY = minYIterate; currY <= maxYIterate; ++currY) {
-                        final int blockY = currY | (currChunkY << 4);
-                        mutablePos.setY(blockY);
-                        for (int currZ = minZIterate; currZ <= maxZIterate; ++currZ) {
-                            final int blockZ = currZ | (currChunkZ << 4);
-                            mutablePos.setZ(blockZ);
-                            for (int currX = minXIterate; currX <= maxXIterate; ++currX) {
-                                final int blockX = currX | (currChunkX << 4);
-                                mutablePos.setX(blockX);
-
-                                final BlockState blockState = blocks.get((currX) | (currZ << 4) | ((currY) << 8));
-
-                                if (((ca.spottedleaf.moonrise.patches.collisions.block.CollisionBlockState)blockState).moonrise$emptyCollisionShape()
-                                    || !blockState.isSuffocating(world, mutablePos)) {
-                                    continue;
-                                }
-
-                                // Yes, it does not use the Entity context stuff.
-                                final VoxelShape collisionShape = blockState.getCollisionShape(world, mutablePos);
-
-                                if (collisionShape.isEmpty()) {
-                                    continue;
-                                }
-
-                                final AABB toCollide = boundingBox.move(-(double)blockX, -(double)blockY, -(double)blockZ);
-
-                                final AABB singleAABB = ((ca.spottedleaf.moonrise.patches.collisions.shape.CollisionVoxelShape)collisionShape).moonrise$getSingleAABBRepresentation();
-                                if (singleAABB != null) {
-                                    if (ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.voxelShapeIntersect(singleAABB, toCollide)) {
-                                        return true;
-                                    }
-                                    continue;
-                                }
-
-                                if (ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.voxelShapeIntersectNoEmpty(collisionShape, toCollide)) {
-                                    return true;
-                                }
-                                continue;
-                            }
-                        }
+        } else {
+            float f = this.dimensions.width() * 0.8F;
+            AABB aabb = AABB.ofSize(this.getEyePosition(), (double)f, 1.0E-6, (double)f);
+            return BlockPos.betweenClosedStream(aabb)
+                .anyMatch(
+                    blockposition -> {
+                        BlockState iblockdata = this.level().getBlockState(blockposition);
+
+                        return !iblockdata.isAir()
+                            && iblockdata.isSuffocating(this.level(), blockposition)
+                            && Shapes.joinIsNotEmpty(iblockdata.getCollisionShape(this.level(), blockposition).move(blockposition), Shapes.create(aabb), BooleanOp.AND);
                     }
-                }
-            }
+                );
         }
-
-        return false;
-        // Paper end - optimise collisions
+        // Paper end - optimise collisions reverted to vanilla
     }
 
     public InteractionResult interact(Player player, InteractionHand hand) {
@@ -4799,126 +4692,79 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         return Mth.lerp(partialTick, this.yRotO, this.yRot);
     }
 
-    // Paper start - optimise collisions
-    public boolean updateFluidHeightAndDoFluidPushing(final TagKey<Fluid> fluid, final double flowScale) {
+    // Paper start - optimise collisions reverted to vanilla
+    public boolean updateFluidHeightAndDoFluidPushing(TagKey<Fluid> fluidTag, double motionScale) {
         if (this.touchingUnloadedChunk()) {
             return false;
-        }
-
-        final AABB boundingBox = this.getBoundingBox().deflate(1.0E-3);
-
-        final Level world = this.level;
-        final int minSection = ca.spottedleaf.moonrise.common.util.WorldUtil.getMinSection(world);
-
-        final int minBlockX = Mth.floor(boundingBox.minX);
-        final int minBlockY = Math.max((minSection << 4), Mth.floor(boundingBox.minY));
-        final int minBlockZ = Mth.floor(boundingBox.minZ);
-
-        // note: bounds are exclusive in Vanilla, so we subtract 1 - our loop expects bounds to be inclusive
-        final int maxBlockX = Mth.ceil(boundingBox.maxX) - 1;
-        final int maxBlockY = Math.min((ca.spottedleaf.moonrise.common.util.WorldUtil.getMaxSection(world) << 4) | 15, Mth.ceil(boundingBox.maxY) - 1);
-        final int maxBlockZ = Mth.ceil(boundingBox.maxZ) - 1;
-
-        final boolean isPushable = this.isPushedByFluid();
-        final BlockPos.MutableBlockPos mutablePos = new BlockPos.MutableBlockPos();
-
-        Vec3 pushVector = Vec3.ZERO;
-        double totalPushes = 0.0;
-        double maxHeightDiff = 0.0;
-        boolean inFluid = false;
-
-        final int minChunkX = minBlockX >> 4;
-        final int maxChunkX = maxBlockX >> 4;
-
-        final int minChunkZ = minBlockZ >> 4;
-        final int maxChunkZ = maxBlockZ >> 4;
-
-        final net.minecraft.world.level.chunk.ChunkSource chunkSource = world.getChunkSource();
-
-        final int chunkLenX = maxChunkX - minChunkX + 1;
-        // chunk index = (x - minX) + (maxX-minX+1)*(z - minZ)
-        //             = x + (maxX-minX+1)*z - (minX + (maxX-minX+1)*minZ)
-        final int chunkOffset = -(minChunkX + chunkLenX*minChunkZ);
-        //             = x + (maxX-minX+1)*z + chunkOffset
-        final net.minecraft.world.level.chunk.LevelChunkSection[][] sections = new net.minecraft.world.level.chunk.LevelChunkSection[chunkLenX * (maxChunkZ - minChunkZ + 1)][];
-
-        // init chunks
-        for (int currChunkZ = minChunkZ; currChunkZ <= maxChunkZ; ++currChunkZ) {
-            for (int currChunkX = minChunkX; currChunkX <= maxChunkX; ++currChunkX) {
-                sections[currChunkX + chunkLenX*currChunkZ + chunkOffset] = chunkSource.getChunk(currChunkX, currChunkZ, net.minecraft.world.level.chunk.status.ChunkStatus.FULL, false).getSections();
-            }
-        }
-
-        for (int currX = minBlockX; currX <= maxBlockX; ++currX) {
-            for (int currY = minBlockY; currY <= maxBlockY; ++currY) {
-                for (int currZ = minBlockZ; currZ <= maxBlockZ; ++currZ) {
-                    final FluidState fluidState = sections[(currX >> 4) + chunkLenX*(currZ >> 4) + chunkOffset][(currY >> 4) - minSection]
-                        .states.get((currX & 15) | ((currZ & 15) << 4) | ((currY & 15) << 8)).getFluidState();
-
-                    if (fluidState.isEmpty() || !fluidState.is(fluid)) {
-                        continue;
-                    }
-
-                    mutablePos.set(currX, currY, currZ);
-
-                    // CraftBukkit start - store last lava contact location
-                    if (fluid == FluidTags.LAVA) {
-                        this.lastLavaContact = mutablePos.immutable();
-                    }
-                    // CraftBukkit end
-
-                    final double height = (double)((float)currY + fluidState.getHeight(world, mutablePos));
-                    final double diff = height - boundingBox.minY;
-
-                    if (diff < 0.0) {
-                        continue;
-                    }
-
-                    inFluid = true;
-                    maxHeightDiff = Math.max(maxHeightDiff, diff);
-
-                    if (!isPushable) {
-                        continue;
-                    }
-
-                    ++totalPushes;
-
-                    final Vec3 flow = fluidState.getFlow(world, mutablePos);
+        } else {
+            AABB aabb = this.getBoundingBox().deflate(0.001);
+            int floor = Mth.floor(aabb.minX);
+            int ceil = Mth.ceil(aabb.maxX);
+            int floor1 = Mth.floor(aabb.minY);
+            int ceil1 = Mth.ceil(aabb.maxY);
+            int floor2 = Mth.floor(aabb.minZ);
+            int ceil2 = Mth.ceil(aabb.maxZ);
+            double d = 0.0;
+            boolean isPushedByFluid = this.isPushedByFluid();
+            boolean flag = false;
+            Vec3 vec3 = Vec3.ZERO;
+            int i = 0;
+            BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos();
+
+            for (int i1 = floor; i1 < ceil; i1++) {
+                for (int i2 = floor1; i2 < ceil1; i2++) {
+                    for (int i3 = floor2; i3 < ceil2; i3++) {
+                        mutableBlockPos.set(i1, i2, i3);
+                        FluidState fluidState = this.level().getFluidState(mutableBlockPos);
+                        if (fluidState.is(fluidTag)) {
+                            double d1 = i2 + fluidState.getHeight(this.level(), mutableBlockPos);
+                            if (d1 >= aabb.minY) {
+                                flag = true;
+                                d = Math.max(d1 - aabb.minY, d);
+                                if (isPushedByFluid) {
+                                    Vec3 flow = fluidState.getFlow(this.level(), mutableBlockPos);
+                                    if (d < 0.4) {
+                                        flow = flow.scale(d);
+                                    }
 
-                    if (maxHeightDiff < 0.4) {
-                        pushVector = pushVector.add(flow.scale(maxHeightDiff));
-                    } else {
-                        pushVector = pushVector.add(flow);
+                                    vec3 = vec3.add(flow);
+                                    i++;
+                                }
+                                // CraftBukkit start - store last lava contact location
+                                if (fluidTag == FluidTags.LAVA) {
+                                    this.lastLavaContact = mutableBlockPos.immutable();
+                                }
+                                // CraftBukkit end
+                            }
+                        }
                     }
                 }
             }
-        }
 
-        this.fluidHeight.put(fluid, maxHeightDiff);
+            if (vec3.length() > 0.0) {
+                if (i > 0) {
+                    vec3 = vec3.scale(1.0 / i);
+                }
 
-        if (pushVector == Vec3.ZERO) {
-            return inFluid;
-        }
+                if (!(this instanceof Player)) {
+                    vec3 = vec3.normalize();
+                }
 
-        // note: totalPushes != 0 as pushVector was changed
-        pushVector = pushVector.scale(1.0 / totalPushes);
-        final Vec3 currMovement = this.getDeltaMovement();
+                Vec3 deltaMovement = this.getDeltaMovement();
+                vec3 = vec3.scale(motionScale);
+                double d2 = 0.003;
+                if (Math.abs(deltaMovement.x) < 0.003 && Math.abs(deltaMovement.z) < 0.003 && vec3.length() < 0.0045000000000000005) {
+                    vec3 = vec3.normalize().scale(0.0045000000000000005);
+                }
 
-        if (!((Entity)(Object)this instanceof Player)) {
-            pushVector = pushVector.normalize();
-        }
+                this.setDeltaMovement(this.getDeltaMovement().add(vec3));
+            }
 
-        pushVector = pushVector.scale(flowScale);
-        if (Math.abs(currMovement.x) < 0.003 && Math.abs(currMovement.z) < 0.003 && pushVector.length() < 0.0045000000000000005) {
-            pushVector = pushVector.normalize().scale(0.0045000000000000005);
+            this.fluidHeight.put(fluidTag, d);
+            return flag;
         }
-
-        this.setDeltaMovement(currMovement.add(pushVector));
-
-        // note: inFluid = true here as pushVector != 0
-        return true;
     }
-    // Paper end - optimise collisions
+    // Paper end - optimise collisions reverted to vanilla
 
     public boolean touchingUnloadedChunk() {
         AABB aabb = this.getBoundingBox().inflate(1.0);
